File: First and Last position of SortedArray.py
""""// Time Complexity : O(log(n))
// Space Complexity :
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :
"""




class Solution:
    def searchRange(self, nums, target):
        l = 0
        r = len(nums) - 1
        ans = [-1, -1]
        while l <= r:
            mid = l + (r - l) // 2
            if nums[mid] == target:
                if mid == 0 or nums[mid] > nums[mid - 1]:
                    ans[0] = mid
                r = mid - 1

            elif nums[mid] < target:
                l = mid + 1
            else:
                r = mid - 1
        l = 0
        r = len(nums) - 1
        while l <= r:
            mid = l + (r - l) // 2
            if nums[mid] == target:
                if mid == len(nums) - 1 or nums[mid] < nums[mid + 1]:
                    ans[1] = mid
                l = mid + 1
                ans[1] = mid
            elif nums[mid] < target:
                l = mid + 1
            else:
                r = mid - 1
        return ans

File: FirstAndLastPositionOfSortedArray.java
//Time Complexity O(log n) space complexity O(1)
class FirstAndLastPositionOfSortedArray {
    public int[] searchRange(int[] nums, int target) {
        if(nums == null && nums.length == 0)
            return new int[] {-1,-1};
        int first = getFirstIndex(nums, target);
        if(first == -1){
            return new int[] {-1,-1};
        }
        int second = getSecondIndex(nums, target, first);

        return new int[] {first,second};

    }
    private int getFirstIndex(int[] nums, int target){
        int low = 0;
        int high = nums.length-1;
        while (low <= high){
            int mid = low + (high - low)/2;
            if(nums[mid]==target){
                if(mid == low || nums[mid-1]!=target)
                    return mid;
                else
                    high = mid - 1;
            }
            else if(target > nums[mid])
                low = mid + 1;
            else
                high = mid -1;

        }
        return  -1;
    }

    private int getSecondIndex(int[] nums, int target, int low){
        int high = nums.length-1;
        while (low<=high){
            int mid = low + (high - low)/2;
            if(nums[mid] == target){
                if(mid == high || nums[mid + 1]!=target)
                    return mid;
                else
                    low = mid + 1;
            }else if(target<nums[mid])
                high = mid - 1;
            else
                low = mid - 1;
        }
        return -1;
    }

}